docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,474 @@
+
+.. code-block:: js
+
+  import { shell } from "execa";
+  shell(cmd); // <-- add 'cmd' as a taint sink
+
+This sink is already recognized by the CodeQL JS analysis, but for the sake of this example we'll show how it could be added as an extension.
+
+This can be achieved with the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["execa", "Member[shell].Argument[0]", "command-line-injection"]
+
+To break this down:
+
+- Since we're adding a new sink, we add a tuple to the **sinkModel** extension point.
+- The first column, **"execa"**, identifies a set of values from which to begin the search for the sink.
+  The string **"execa"** means we start at the places where the codebase imports the NPM package **execa**.
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+
+  - **Member[shell]** selects accesses to the **shell** member of the **execa** package.
+  - **Argument[0]** selects the first argument to calls to that member.
+
+- **command-line-injection** indicates that this is considered a sink for the command injection query.
+
+Example: Taint sources from window 'message' events
+---------------------------------------------------
+
+In this example, we'll show how the **event.data** expression below could be marked as a remote flow source:
+
+.. code-block:: js
+
+  window.addEventListener("message", function (event) {
+    let data = event.data; // <-- add 'event.data' as a taint source
+  });
+
+This source is already known by the CodeQL JS analysis, but we'll show how it could be added as an extension. This can be achieved with the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sourceModel
+      data:
+        - [
+            "global",
+            "Member[addEventListener].Argument[1].Parameter[0].Member[data]",
+            "remote",
+          ]
+
+To break this down:
+
+- Since we're adding a new taint source, we add a tuple to the **sourceModel** extension point.
+- The first column, **"global"**, begins the search at references to the global object (also known as **window** in browser contexts). This is a special JavaScript object that contains all global variables and methods.
+- **Member[addEventListener]** selects accesses to the **addEventListener** member.
+- **Argument[1]** selects the second argument of calls to that member (the argument containing the callback).
+- **Parameter[0]** selects the first parameter of the callback (the parameter named **event**).
+- **Member[data]** selects accesses to the **data** property of the event object.
+- Finally, the kind **remote** indicates that this is considered a source of remote flow.
+
+Continued example: Restricting the event type
+---------------------------------------------
```

Comment:
I'd recommend using RST subsection formatting here:

```suggestion
Restricting the event type
~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Hmm, would you consider using bold in that case? I'm just thinking because otherwise it looks like it's on the same level as other 'main' examples. Given this is only for internal Hubber use initially, it's not vital that we resolve it now either way, so I'll leave to you.

Reply:
> I'm just thinking because otherwise it looks like it's on the same level as other 'main' examples.

I have no problem with this. There is one section per example, and it's clearly indicated when an example builds onto the previous one.